File: shared-native/bindings/bun/CompleteTransaction.ts
import { ptr, toArrayBuffer } from 'bun:ffi'
import type { ErrorDetails, GrdtTransactionType, VoidResult } from '../../'
import { LedgerAnchor } from './LedgerAnchor'
import { blockchain_core, SIGN_PUBLIC_KEY_SIZE } from './library'
import { grddTimestampToDate } from './utils'

const registry = new FinalizationRegistry((handle: bigint) => {
  if (handle && blockchain_core.symbols.grdr_complete_transaction_release) {
    blockchain_core.symbols.grdr_complete_transaction_free(handle)
    handle = 0n
  }
})

export class NativeCompleteTransaction {
  private handle: bigint | null = null

  public constructor() {
    // call internally malloc so this.handle is a number, containing ptr address
    this.handle = blockchain_core.symbols.grdr_complete_transaction_create()
    if (!this.handle) {
      throw new Error('Failed to create transaction')
    }
    // auto clean up
    registry.register(this, this.handle)
  }

  public initFromProtobuf(serialized: Uint8Array, communityUuid: Uint8Array | string): VoidResult {
    if (!this.handle) {
      throw new Error('[CompleteTransaction.initFromProtobuf] Object not initalized')
    }
    let communityUuidBuffer: Uint8Array
    if (typeof communityUuid === 'string') {
      communityUuidBuffer = new Uint8Array(Buffer.from(communityUuid.replace(/-/g, ''), 'hex'))
    } else {
      communityUuidBuffer = communityUuid
    }
    const communityUuidPtr = ptr(communityUuidBuffer)
    const serializedPtr = ptr(serialized)
    let bufferSize = 4096
    let resultString = 'GRD_SUCCESS'
    do {
      if (bufferSize >= 1024 * 1024) {
        return {
          success: false,
          error: new Error('[CompleteTransaction.initFromProtobuf] serialized data is to big'),
        }
      }
      const resultBuffer = new Uint8Array(bufferSize)
      const result = blockchain_core.symbols.grdr_complete_transaction_init_from_protobuf(
        this.handle,
        serializedPtr,
        serialized.byteLength,
        communityUuidPtr,
        resultBuffer,
        bufferSize,
      )
      resultString = blockchain_core.symbols.grd_result_to_string(result).toString()
      bufferSize *= 2
    } while (
      resultString === 'GRD_ERROR_OUT_OF_MEMORY' ||
      resultString === 'GRD_ERROR_DESTINATION_BUFFER_TO_SMALL'
    )
    if (resultString !== 'GRD_SUCCESS') {
      return {
        success: false,
        error: new Error(
          `[CompleteTransaction.initFromProtobuf] Couldn't parse CompleteTransaction from serialized, returned: ${resultString}`,
        ),
      }
    }
    return { success: true }
  }
  public validate(verifySignatures: boolean = true): VoidResult<ErrorDetails> {
    if (!this.handle) {
      throw new Error('[CompleteTransaction.validate] Object not initalized')
    }

    const errorDetailsPtr = blockchain_core.symbols.grd_error_details_create(null)
    const result = blockchain_core.symbols.grdi_validate_complete_transaction_flat_options(
      this.handle,
      verifySignatures,
      errorDetailsPtr,
    )
    const resultString = blockchain_core.symbols.grdi_validate_result_to_string(result)
    let errorDetails: ErrorDetails | undefined
    if (resultString.toString() !== 'GRDI_VALIDATE_SUCCESS') {
      const message = blockchain_core.symbols.grd_error_details_get_message(errorDetailsPtr)
      const actual = blockchain_core.symbols.grd_error_details_get_message(errorDetailsPtr)
      const expected = blockchain_core.symbols.grd_error_details_get_message(errorDetailsPtr)
      errorDetails = {
        name: resultString,
        message,
        actual,
        expected,
      }
    }
    blockchain_core.symbols.grd_error_details_free(errorDetailsPtr)
    return errorDetails ? { success: false, error: errorDetails } : { success: true }
  }

  public getConfirmedAt(): Date {
    if (!this.handle) {
      throw new Error('[CompleteTransaction.getConfirmedAt] Object not initalized')
    }
    const timestampHandle = blockchain_core.symbols.grdr_complete_transaction_get_confirmed_at(
      this.handle,
    )
    return grddTimestampToDate(timestampHandle)
  }

  public getCreatedAt(): Date {
    if (!this.handle) {
      throw new Error('[CompleteTransaction.getCreatedAt] Object not initalized')
    }
    const timestampHandle = blockchain_core.symbols.grdr_complete_transaction_get_created_at(
      this.handle,
    )
    return grddTimestampToDate(timestampHandle)
  }

  public getLedgerAnchor(): LedgerAnchor {
    if (!this.handle) {
      throw new Error('[CompleteTransaction.getLedgerAnchor] Object not initalized')
    }
    return LedgerAnchor.copy(
      blockchain_core.symbols.grdr_complete_transaction_get_ledger_anchor(this.handle),
    )
  }

  public getSenderPublicKey(): Uint8Array | null {
    if (!this.handle) {
      throw new Error('[CompleteTransaction.getSenderPublicKey] Object not initalized')
    }

    const keyPtr = blockchain_core.symbols.grdr_complete_transaction_get_sender_public_key(
      this.handle,
    )
    if (!keyPtr) {
      return null
    }
    return new Uint8Array(toArrayBuffer(keyPtr, 0, SIGN_PUBLIC_KEY_SIZE), 0, SIGN_PUBLIC_KEY_SIZE)
  }

  public getRecipientPublicKey(): Uint8Array | null {
    if (!this.handle) {
      throw new Error('[CompleteTransaction.getRecipientPublicKey] Object not initalized')
    }

    const keyPtr = blockchain_core.symbols.grdr_complete_transaction_get_recipient_public_key(
      this.handle,
    )
    if (!keyPtr) {
      return null
    }
    return new Uint8Array(toArrayBuffer(keyPtr, 0, SIGN_PUBLIC_KEY_SIZE), 0, SIGN_PUBLIC_KEY_SIZE)
  }

  public getSenderCommunityUuid(): string | null {
    if (!this.handle) {
      throw new Error('[CompleteTransaction.getSenderCommunityUuid] Object not initalized')
    }

    const uuidPtr = blockchain_core.symbols.grdr_complete_transaction_get_sender_community_uuid(
      this.handle,
    )
    if (!uuidPtr) {
      return null
    }
    const resultBuffer = new Uint8Array(37)
    const resultBufferPtr = ptr(resultBuffer)
    blockchain_core.symbols.grdu_uuid_to_string(resultBufferPtr, uuidPtr)
    return Buffer.from(resultBuffer).toString('utf8').slice(0, 36)
  }

  public getRecipientCommunityUuid(): string | null {
    if (!this.handle) {
      throw new Error('[CompleteTransaction.getRecipientCommunityUuid] Object not initalized')
    }

    const uuidPtr = blockchain_core.symbols.grdr_complete_transaction_get_recipient_community_uuid(
      this.handle,
    )
    if (!uuidPtr) {
      return null
    }
    const resultBuffer = new Uint8Array(37)
    const resultBufferPtr = ptr(resultBuffer)
    blockchain_core.symbols.grdu_uuid_to_string(resultBufferPtr, uuidPtr)
    return Buffer.from(resultBuffer).toString('utf8').slice(0, 36)
  }

  public getRegisteredAccount(): Uint8Array | null {
    if (!this.handle) {
      throw new Error('[CompleteTransaction.getRegisteredAccount] Object not initalized')
    }

    const accountPtr = blockchain_core.symbols.grdr_complete_transaction_get_registered_account(
      this.handle,
    )
    if (!accountPtr) {
      return null
    }
    return new Uint8Array(
      toArrayBuffer(accountPtr, 0, SIGN_PUBLIC_KEY_SIZE),
      0,
      SIGN_PUBLIC_KEY_SIZE,
    )
  }

  public getAmount(): bigint {
    if (!this.handle) {
      throw new Error('[CompleteTransaction.getAmount] Object not initalized')
    }

    return blockchain_core.symbols.grdr_complete_transaction_get_amount(this.handle)
  }

  public getAccountBalanceForPublicKey(
    publicKey: Uint8Array | string,
  ): { balance: bigint; coinCommunityUuid: string } | null {
    if (!this.handle) {
      throw new Error('[CompleteTransaction.getAccountBalanceForPublicKey] Object not initalized')
    }

    let publicKeyBuffer: Uint8Array
    if (typeof publicKey === 'string') {
      if (publicKey.length !== SIGN_PUBLIC_KEY_SIZE * 2) {
        throw new Error(
          `[CompleteTransaction.getAccountBalanceForPublicKey] Expected publicKey to be size ${SIGN_PUBLIC_KEY_SIZE * 2} as string`,
        )
      }
      publicKeyBuffer = new Uint8Array(Buffer.from(publicKey, 'hex'))
      if (publicKeyBuffer.length !== SIGN_PUBLIC_KEY_SIZE) {
        throw new Error(
          '[CompleteTransaction.getAccountBalanceForPublicKey] Expected publicKey to be valid hex string',
        )
      }
    } else {
      if (publicKey.length !== SIGN_PUBLIC_KEY_SIZE) {
        throw new Error(
          `[CompleteTransaction.getAccountBalanceForPublicKey] Expected publicKey to be size ${SIGN_PUBLIC_KEY_SIZE} as Uint8Array`,
        )
      }
      publicKeyBuffer = publicKey
    }

    const publicKeyPtr = ptr(publicKeyBuffer)
    const result = blockchain_core.symbols.transaction_get_account_balance_for_public_key(
      this.handle,
      publicKeyPtr,
    )

    if (result === 0) {
      return null
    }

    const communityUuidPtr = blockchain_core.symbols.grdw_account_balance_get_community_uuid(result)
    const communityUuidStringBuffer = new Uint8Array(37)
    const communityUuidStringBufferPtr = ptr(communityUuidStringBuffer)
    blockchain_core.symbols.grdu_uuid_to_string(communityUuidStringBufferPtr, communityUuidPtr)

    return {
      balance: BigInt(blockchain_core.symbols.grdw_account_balance_get_balance(result)),
      coinCommunityUuid: Buffer.from(communityUuidStringBuffer).toString('utf8').slice(0, 36),
    }
  }

  public getTransactionType(): GrdtTransactionType {
    if (!this.handle) {
      throw new Error('[CompleteTransaction.getTransactionType] Object not initalized')
    }

    return blockchain_core.symbols
      .grdt_transaction_to_string(
        blockchain_core.symbols.grdr_complete_transaction_get_transaction_type(this.handle),
      )
      .toString()
  }

  public getTargetDate(): Date | null {
    if (!this.handle) {
      throw new Error('[CompleteTransaction.getTargetDate] Object not initalized')
    }

    const timestampSeconds = blockchain_core.symbols.grdr_complete_transaction_get_target_date(
      this.handle,
    )
    if (!timestampSeconds) {
      return null
    }
    return new Date(Number(timestampSeconds * 1000n))
  }

  public getTimeoutDuration(): bigint {
    if (!this.handle) {
      throw new Error('[CompleteTransaction.getTimeoutDuration] Object not initalized')
    }

    return blockchain_core.symbols.grdr_complete_transaction_get_timeout_duration(this.handle)
  }
}

File: shared-native/bindings/napi/CompleteTransaction.cpp
#include "CompleteTransaction.h"

#include "gradido_blockchain_core/const.h"
#include "gradido_blockchain_core/data/runtime/complete_transaction.h"
#include "gradido_blockchain_core/error_details.h"
#include "gradido_blockchain_core/interactions/validate/context.h"
#include "gradido_blockchain_core/interactions/validate/result_type.h"
#include "gradido_blockchain_core/interactions/validate/options.h"
#include "gradido_blockchain_core/result.h"
#include "gradido_blockchain_core/utils/converter.h"
#include "napi.h"

namespace gradido::data::runtime {
    
    Napi::Object CompleteTransaction::Init(Napi::Env env, Napi::Object exports) {
        Napi::Function func = DefineClass(env, "CompleteTransaction", {
            InstanceMethod("initFromProtobuf", &CompleteTransaction::InitFromProtobuf),
            InstanceMethod("validate", &CompleteTransaction::Validate),
            InstanceMethod("getAccountBalanceForPublicKey", &CompleteTransaction::GetAccountBalanceForPublicKey),
            InstanceMethod("getSenderPublicKey", &CompleteTransaction::GetSenderPublicKey),
            InstanceMethod("getRecipientPublicKey", &CompleteTransaction::GetRecipientPublicKey),
            InstanceMethod("getSenderCommunityUuid", &CompleteTransaction::GetSenderCommunityUuid),
            InstanceMethod("getRecipientCommunityUuid", &CompleteTransaction::GetRecipientCommunityUuid),
            InstanceMethod("getRegisteredAccount", &CompleteTransaction::GetRegisteredAccount),
            InstanceMethod("getAmount", &CompleteTransaction::GetAmount),
            InstanceMethod("getTransactionType", &CompleteTransaction::GetTransactionType),
            InstanceMethod("getTargetDate", &CompleteTransaction::GetTargetDate),
            InstanceMethod("getTimeoutDuration", &CompleteTransaction::GetTimeoutDuration)
        });
        
        exports.Set("CompleteTransaction", func);
        return exports;
    }
    
    CompleteTransaction::CompleteTransaction(const Napi::CallbackInfo& info) 
        : Napi::ObjectWrap<CompleteTransaction>(info) 
    {   
        grdr_complete_transaction_init(&m_tx);   
    }
    
    CompleteTransaction::~CompleteTransaction() {
        grdr_complete_transaction_release(&m_tx);
    }

    Napi::Value CompleteTransaction::InitFromProtobuf(const Napi::CallbackInfo& info)
    {
        auto env = info.Env();
                
        if (info.Length() < 2) {
            throw Napi::Error::New(info.Env(), "Expected two arguments: serialized Transaction (UInt8Array) and community uuid (UInt8Array(16) or string(36))");
            return env.Undefined();
        }
        if (!info[0].IsBuffer()) {
            Napi::TypeError::New(env, "Expected first argument to be a Uint8Array").ThrowAsJavaScriptException();
            return env.Undefined();
        }
        uint8_t communityUuid[UUID_BINARY_SIZE];
        if (info[1].IsBuffer()) {
            Napi::Buffer<uint8_t> communityUuidBuffer = info[1].As<Napi::Buffer<uint8_t>>();
            if (communityUuidBuffer.Length() != UUID_BINARY_SIZE) {
                Napi::TypeError::New(env, "Expected second argument to be size 16 as Uint8Array").ThrowAsJavaScriptException();
                return env.Undefined();
            }
            memcpy(communityUuid, communityUuidBuffer.Data(), UUID_BINARY_SIZE);
        } else if (info[1].IsString()) {
            auto communityUuidString = info[1].As<Napi::String>().Utf8Value();            
            if (communityUuidString.size() != 36) {
                Napi::TypeError::New(env, "Expected second argument to be size 36 as string").ThrowAsJavaScriptException();
                return env.Undefined();
            }
            grd_result result = grdu_uuid_from_string(communityUuid, communityUuidString.c_str());
            if (GRD_SUCCESS != result) {
                Napi::TypeError::New(env, "Expected second argument to be valid uuid string").ThrowAsJavaScriptException();
                return env.Undefined();
            }
        } else {
            Napi::TypeError::New(env, "Expected second argument to be a Uint8Array(16) or string(36)").ThrowAsJavaScriptException();
            return env.Undefined();
        }
        Napi::Buffer<uint8_t> serializedTx = info[0].As<Napi::Buffer<uint8_t>>();

        int bufferSize = 4096;
        uint8_t* dynBuffer = nullptr;
        uint8_t buffer[bufferSize];
        grd_result init_result = grdr_complete_transaction_init_from_protobuf(
            &m_tx, 
            serializedTx.Data(), serializedTx.Length(), 
            communityUuid, 
            buffer, bufferSize
        );
        
        while(GRD_ERROR_OUT_OF_MEMORY == init_result || GRD_ERROR_DESTINATION_BUFFER_TO_SMALL == init_result) {
            bufferSize *= 2;
            // 1 MB should be more as enough
            if(bufferSize >= 1024 * 1024 * 1024) { break;}
            dynBuffer = (uint8_t*)malloc(bufferSize);
            init_result = grdr_complete_transaction_init_from_protobuf(
                &m_tx, 
                serializedTx.Data(), serializedTx.Length(), 
                communityUuid, 
                dynBuffer, bufferSize
            );
            free(dynBuffer);
        };
        Napi::Object result = Napi::Object::New(env);        
        if (GRD_SUCCESS != init_result) {
            std::string message = "deserialize or mapping failed: ";
            message += grd_result_to_string(init_result);
            Napi::Object error = Napi::Object::New(env);
            error.Set("name", Napi::String::New(env, grd_result_to_string(init_result)));
            error.Set("message", Napi::String::New(env, "Deserialize or mapping failed"));
            result.Set("success", Napi::Boolean::New(env, false));    
            result.Set("error", error);
            return result;
        }
        result.Set("success", Napi::Boolean::New(env, true));    
        return result;
    }

    Napi::Value CompleteTransaction::Validate(const Napi::CallbackInfo& info)
    {
        auto env = info.Env();
        bool verifySignatures = true;
        if (info.Length() >= 1) {
            if (info[0].IsBoolean()) {
                verifySignatures = info[0].As<Napi::Boolean>();
            }
        }
        
        grdi_validate_options opt = {
          .enable_verify = verifySignatures
        };
        grd_error_details errorDetails;
        uint8_t errorStringBuffer[256];
        grd_memory alloc;
        grd_memory_init_arena_static(&alloc, errorStringBuffer, 256);
        // error details will use malloc for error message, when alloc has run out of memory
        grd_result errorDetailsInitResult = grd_error_details_init(&errorDetails, &alloc);
        if (errorDetailsInitResult != GRD_SUCCESS) { 
            std::string message = "Error on error details init: ";
            message += grd_result_to_string(errorDetailsInitResult);
            Napi::TypeError::New(env, message.c_str()).ThrowAsJavaScriptException();
            return env.Undefined();
        }
        grdi_validate_result_type validateResult = grdi_validate_complete_transaction(&m_tx, &opt, &errorDetails);
        Napi::Object result = Napi::Object::New(env);
        if (validateResult != GRDI_VALIDATE_SUCCESS) {
            result.Set("success", Napi::Boolean::New(env, false));
            Napi::Object error = Napi::Object::New(env);
            error.Set("name", Napi::String::New(env, grdi_validate_result_to_string(validateResult)));
            if (errorDetails.message) {
                error.Set("message", Napi::String::New(env, errorDetails.message));
            }
            if (errorDetails.actual) {
                error.Set("actual", Napi::String::New(env, errorDetails.actual));
            }
            if (errorDetails.expected) {
                error.Set("expected", Napi::String::New(env, errorDetails.expected));
            }
            result.Set("error", error);        
        } else {
            result.Set("success", Napi::Boolean::New(env, true));
        }
        grd_error_details_release(&errorDetails);
        return result;
    }
    
    Napi::Value CompleteTransaction::GetSenderPublicKey(const Napi::CallbackInfo& info) {
        auto env = info.Env();
        const uint8_t* key = grdr_complete_transaction_get_sender_public_key(&m_tx);
        if (!key) return env.Null();
        return Napi::Buffer<uint8_t>::Copy(env, key, SIGN_PUBLIC_KEY_SIZE);
    }

    Napi::Value CompleteTransaction::GetRecipientPublicKey(const Napi::CallbackInfo& info)
    {
        auto env = info.Env();
        const uint8_t* key = grdr_complete_transaction_get_recipient_public_key(&m_tx);
        if (!key) return env.Null();
        return Napi::Buffer<uint8_t>::Copy(env, key, SIGN_PUBLIC_KEY_SIZE);
    }

    Napi::Value CompleteTransaction::GetSenderCommunityUuid(const Napi::CallbackInfo& info) {
        auto env = info.Env();
        const uint8_t* key = grdr_complete_transaction_get_sender_community_uuid(&m_tx);
        if (!key) return env.Null();
        
        char buffer[37];
        grdu_uuid_to_string(buffer, key);        
        return Napi::String::New(env, buffer);
    }
    
    Napi::Value CompleteTransaction::GetRecipientCommunityUuid(const Napi::CallbackInfo& info) {
        auto env = info.Env();
        const uint8_t* key = grdr_complete_transaction_get_recipient_community_uuid(&m_tx);
        if (!key) return env.Null();
        
        char buffer[37];
        grdu_uuid_to_string(buffer, key);
        return Napi::String::New(env, buffer);
    }

    Napi::Value CompleteTransaction::GetRegisteredAccount(const Napi::CallbackInfo& info)
    {
        auto env = info.Env();
        const uint8_t* key = grdr_complete_transaction_get_registered_account(&m_tx);
        if (!key) return env.Null();
        return Napi::Buffer<uint8_t>::Copy(env, key, SIGN_PUBLIC_KEY_SIZE);
    }
    
    Napi::Value CompleteTransaction::GetAccountBalanceForPublicKey(const Napi::CallbackInfo& info)
    {
        auto env = info.Env();
        if (info.Length() < 1) {
            throw Napi::Error::New(info.Env(), "Expected one argument: publicKey as Uint8Array(32) or hex string (64)");
            return env.Undefined();
        }
        uint8_t publicKey[SIGN_PUBLIC_KEY_SIZE];
        if (info[0].IsBuffer()) {
            Napi::Buffer<uint8_t> publicKeyBuffer = info[0].As<Napi::Buffer<uint8_t>>();
            if (publicKeyBuffer.Length() != SIGN_PUBLIC_KEY_SIZE) {
                Napi::TypeError::New(env, "Expected first argument to be size 32 as Uint8Array").ThrowAsJavaScriptException();
                return env.Null();
            }
            memcpy(publicKey, publicKeyBuffer.Data(), SIGN_PUBLIC_KEY_SIZE);
        } else if (info[0].IsString()) {
            auto publicKeyString = info[0].As<Napi::String>().Utf8Value();            
            if (publicKeyString.size() != SIGN_PUBLIC_KEY_SIZE * 2) {
                Napi::TypeError::New(env, "Expected first argument to be size 64 as string").ThrowAsJavaScriptException();
                return env.Null();
            }
            grd_result result = grdu_binary_from_hex(publicKey, publicKeyString.c_str());
            if (GRD_SUCCESS != result) {
                Napi::TypeError::New(env, "Expected first argument to be valid hex string").ThrowAsJavaScriptException();
                return env.Null();
            }
        } else {
            Napi::TypeError::New(env, "Expected first argument to be a Uint8Array(32) or hex string (64)").ThrowAsJavaScriptException();
            return env.Null();
        }
        const grdw_account_balance* account_balance = grdr_complete_transaction_get_account_balance_for_public_key(&m_tx, publicKey);
        if (!account_balance) {
            return env.Null();
        }
        
        Napi::Object result = Napi::Object::New(env);
        result.Set("balance", Napi::Number::New(env, grdw_account_balance_get_balance(account_balance)));
        result.Set("publicKey", Napi::Buffer<uint8_t>::Copy(env, grdw_account_balance_get_public_key(account_balance), SIGN_PUBLIC_KEY_SIZE));        
        // Community UUID as string
        char uuidString[37];
        grdu_uuid_to_string(uuidString, grdw_account_balance_get_community_uuid(account_balance));
        result.Set("communityUuid", Napi::String::New(env, uuidString));
        
        return result;
    }
    
    Napi::Value CompleteTransaction::GetTransactionType(const Napi::CallbackInfo& info) {
        return Napi::String::New(info.Env(), grdt_transaction_to_string(grdr_complete_transaction_get_transaction_type(&m_tx)));
    }
    
    Napi::Value CompleteTransaction::GetAmount(const Napi::CallbackInfo& info) {
        return Napi::BigInt::New(info.Env(), grdr_complete_transaction_get_amount(&m_tx));
    }

    Napi::Value CompleteTransaction::GetTargetDate(const Napi::CallbackInfo& info) {
        auto env = info.Env();
        auto timestampSeconds = grdr_complete_transaction_get_target_date(&m_tx);
        if (timestampSeconds <= 0) {
            return env.Null();
        }
        return Napi::Date::New(env, static_cast<double>(timestampSeconds) * 1000.0);
    }
    Napi::Value CompleteTransaction::GetTimeoutDuration(const Napi::CallbackInfo& info) {
        return Napi::BigInt::New(info.Env(), grdr_complete_transaction_get_timeout_duration(&m_tx));
    }
}